### ISSUE TITLE
local path dependencies behaviour broken

### ISSUE THREAD

**Issue by wizardofeast** [npm/npm#17615]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [X] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?
We have local package A depends npm package B. If another local package C depends A, npm install under C causes:
1. npm installs all dependencies under A
2. copies everything from A into C node_modules neglecting any ignore pattern under A such as "files" in package.json.
3. do not apply "dependencies will be added as close to the top as is possible" which results misplaced dependencies.

#### How can the CLI team reproduce the problem?
A package.json:

{
  "version": "1.0.0",
  "name": "A",
  "title": "A",
  "description": "A",
  "dependencies": { "d3": "^4.9.1" }
  "files": {"output"}
}

C package.json 
{
  "version": "1.0.0",
  "name": "C",
  "title": "C",
  "description": "C",
  "dependencies": { "A": "file:../A }
}
under C : npm install 

### supporting information:
 - `npm -v` prints: 5.0.4
 - `node -v` prints: v8.1.3
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: Windows
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by wizardofeast** [comment]
Previously all the packages are copied to .staging whether remote or local now only remote are copied.

**Comment by wizardofeast** [comment]
https://github.com/npm/npm/pull/15900

**Comment by wizardofeast** [comment]
Solved the problem via npm pack A and referencing to pack file (tgz).

**Comment by piniu** [comment]
We are experience same issue where dependencies are not downloaded(for very certain packages). The only work around we found is to delete the node_module and then `npm install` . 
Can you explain what you did? we could not figure any solution from #15900

**Comment by wizardofeast** [comment]
Due to the install process in https://github.com/npm/npm/pull/15900 , local dependencies behave exactly that I have described above. 
So I packed the package A with "npm pack" which results lib-1.0.0.tgz file which is also filtered out unnecessary files such as node_modules. Then changed the dependency value to "A":"file://../A/output/lib-1.0.0.tgz" in C. 
It was so frustrating cause we worked with multiple local packages in our team and after the update (node.js) npm updated to 5.0.* and everything gone missing.
